File: 63-unique-paths-ii/unique-paths-ii.js
/**
 * @param {number[][]} obstacleGrid
 * @return {number}
 */
var uniquePathsWithObstacles = function(obstacleGrid) {
    let m = obstacleGrid.length;
    let n = obstacleGrid[0].length;
    let dp = Array.from({length:m},()=>new Array(n).fill(0));
    let flag = false //flag remembers whether an obstacle has already blocked the path, so every cell beyond that obstacle in the first row/column gets 0
    for(let i=0;i<m;i++) {
        if(flag || obstacleGrid[i][0]){
            flag = true
            dp[i][0] = 0;
        }else{
            dp[i][0] = 1;
        }
    }
    flag = false;
    for(let j=0;j<n;j++) {
        if(flag || obstacleGrid[0][j]){
            flag = true
            dp[0][j] = 0;
        }else{
            dp[0][j] = 1;
        }
    }
    for(let i=1;i<m;i++){
        for(let j=1;j<n;j++) {
            if(obstacleGrid[i][j] === 1){
                dp[i][j] = 0;
            }else{
                dp[i][j] = dp[i-1][j] + dp[i][j-1];
            }
            
        }
    }
    return dp[m-1][n-1];
};

File: 63-unique-paths-ii/unique-paths-ii.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {

        int row = obstacleGrid.size();
        int col = obstacleGrid[0].size();

        vector<vector<long long>> dp(row, vector<long long>(col, 0));

        bool lionSeenLastCol = false;

        for(int i=row-1;i>=0;i--){

            //1.Is there lion in cell?
            //2.Is there any lion in my path ahead

            if(obstacleGrid[i][col-1]==0 and lionSeenLastCol==false)
            {
                dp[i][col-1]=1;
            }else{
                lionSeenLastCol=true;
                dp[i][col-1]=0;
            }
        }

        bool lionSeenLastRow = false;

        for(int j=col-1;j>=0;j--){

            //1.Is there lion in cell?
            //2.Is there any lion in my path ahead

            if(obstacleGrid[row-1][j]==0 and lionSeenLastRow==false)
            {
                dp[row-1][j]=1;
            }else{
                lionSeenLastRow=true;
                dp[row-1][j]=0;
            }
        }

        for(int r=row-2;r>=0;r--){
            for(int c=col-2;c>=0;c--){
                if(obstacleGrid[r][c]!=1){
                    long long waysFromRight=dp[r][c+1];
                    long long waysFromDown=dp[r+1][c];

                    dp[r][c]=waysFromRight+waysFromDown;
                }
            }
        }

        return dp[0][0];
    }
};